io-sim/src/Control/Monad/IOSim/Types.hs

```diff
@@ -322,6 +323,32 @@ instance MonadThrow (STM s) where
 instance Exceptions.MonadThrow (STM s) where
   throwM = MonadThrow.throwIO
 
+
+instance MonadCatch (STM s) where
+
+  catch action handler = STM $ oneShot $ \k -> CatchStm (runSTM action) (runSTM . fromHandler handler) k
+    where
+      -- Get a total handler from the given handler
+      fromHandler :: Exception e => (e -> STM s a) -> SomeException -> STM s a
+      fromHandler h e = case fromException e of
+        Nothing -> throwIO e  -- Rethrow the exception if handler does not handle it.
+        Just e' -> h e'
+
+  -- STM actions are always run inside `execAtomically` and behave as if masked
+  -- Another point to note that the default implementation of `generalBracket` needs
+  -- mask, and is part of `MonadThrow`. For STM, it probably does not make sense to have
+  -- masking.
```

Comment:
```suggestion
  -- mask, and is part of `MonadThrow`. For STM, we don't need masking because
  -- async exceptions are handled outside of `execAtomically`.
```